InfiniteRecharge/src/main/java/frc/robot/subsystem/navigation/systems/RobotSystem1.java

```diff
@@ -0,0 +1,149 @@
+                tempVs.set(1, vR);
+
+
+
+                double VL = drive.getLeftVolts();
+                double VR = drive.getRightVolts();
+
+                tempUs.set(0, VL);
+                tempUs.set(1, VR);
+                
+                SimpleMatrix as = config.drive.Ac_char.mult(tempVs).plus(config.drive.Bc_char.mult(tempUs));
+
+                xDot[X_X] = v * Math.cos(theta);
+                xDot[X_Y] = v * Math.sin(theta);
+                xDot[X_THETA] = (vR - vL) / (config.drive.trackWidth_in * NavigationConstants.IN_TO_M);
+                xDot[X_VL] = as.get(0);
+                xDot[X_VR] = as.get(1);
+                xDot[X_OFFSET] = 0; // never changes (theoretically). In reality, IMU drifts a bit so it's okay if it changes
+            }
+        };
+    }
+
+    public static OutputObserver generateObserver(Config config, ShooterSubsystem shooter, VisionSubsystem vision) {
+        return new OutputObserver(6) {
+            @Override
+            public SimpleMatrix getOutput(SimpleMatrix state, SimpleMatrix input, double t, int k) {
+                double theta = state.get(X_THETA);
+                double azimuth = Math.toRadians(shooter.getAzimuthDeg());
+
+                double x = state.get(X_X)
+                    + config.shooter.xTurretCenter * Math.cos(theta)
+                    - config.shooter.yTurretCenter * Math.sin(theta) // center of turret doesn't rotate with turret, only robot
+                    + config.shooter.rLL * Math.cos(theta + azimuth);
+                
+                double y = state.get(X_Y)
+                    + config.shooter.xTurretCenter * Math.sin(theta)
+                    + config.shooter.yTurretCenter * Math.cos(theta) // center of turret doesn't rotate with turret, only robot
+                    + config.shooter.rLL * Math.sin(theta + azimuth);
+                
+                // calcualte distance to target and convert to inches
+                double d = Math.sqrt(
+                    Math.pow(x - NavigationConstants.TARGET_X, 2)
+                    + Math.pow(y - NavigationConstants.TARGET_Y, 2)
+                ) / NavigationConstants.IN_TO_M;
+
+                double tx = Math.atan2(NavigationConstants.TARGET_Y - y, NavigationConstants.TARGET_X - x);
+
+
+
+                double[] out = new double[Y_LENGTH];
+
+                out[Y_TX] = tx;
+                out[Y_TY] = vision.approximateTyFromDistance(d);
+                out[Y_VL] = state.get(X_VL);
+                out[Y_VR] = state.get(X_VR);
+                out[Y_YAW] = theta + state.get(X_OFFSET);
+
+                return new SimpleMatrix(Y_LENGTH, 1, true, out);
+            }
+        };
+    }
+
+
+
+    public RobotSystem1(Config config, NavigationSubsystem navigation) {
+        super(
+            generateModel(navigation.getConfig(), navigation.getDrive()),
+            generateObserver(config, navigation.getShooter(), navigation.getVision())
+        );
+    }
```

Comment:
This is going to be hard to track. The RobotSystem requires the navigation subsystem so it can get the DriveSubsystem, VisionSubsystem. and shooterSubsystem....

If it depends on those subsystems, it should just include those subsystems as constructor args, rather than implicitly including them with the navigation subsystem. 

If the RobotSystem requires those subsystems it should explicitly require them.

Reply:
It would be easier to test this class and it would be more shareable if it didn't depend on subsystems at all, but instead took a series of functions as arguments like

```java
public RobotSystem(Config config, DoubleSupplier leftVoltSupplier, DoubleSupplier rightVoltSupplier, DoubleSupplier azimuthSupplier, DoubleSupplier distanceSupplier) {
```

This way our real robot supplies the values when running, but for testing we can create our own mock suppliers that supply whatever values we want so we can ensure the test works without having to mock out entire subsystems. 

The config is still messy. There is probably something similar we could do... Maybe create a separate config object with just the values this RobotSystem needs?

Reply:
yeah it was a very messy solution. I'll put suppliers in